Importing Windows PowerShell modules in PowerShell Core

anmenaga opened the pull request:
RFC about Importing Windows PowerShell modules in PowerShell Core.

iSazonov commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
```diff
@@ -0,0 +1,106 @@
+PS C:\> $PSVersionTable.PSEdition
+Core
+PS C:\> Import-Module DesktopOnlyModuleOnSystem32ModulePath
+VERBOSE: Please note that module 'DesktopOnlyModuleOnSystem32ModulePath' is imported in a remote Windows PowerShell session, so all objects that are returned by commands from this module are deserialized and are not 'live' objects.
```
I am not sure that user scripts will works well with deserialized objects. I always had problems when I switched to deserialized objects. I'd expect that we use the technology only for modules compatible with serialization.

joeyaiello replied:
Unfortunately, I think it's impossible for us to know whether that's the case. I still think we're going to end up with more successful script executions if we automatically load system32 modules (which are *known* to be incompatible post-Server 2019 and are demarcated as such) than if we simply fail to load them

iSazonov commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
```diff
@@ -0,0 +1,106 @@
+    I can use modules written for Windows PowerShell in PowerShell Core,
+    so that I could migrate from Windows PowerShell to PowerShell Core.
+
+## User Experience
```
Please explicitly say where is _current_ experience and _new_ expirience.

anmenaga replied:
fixed.

iSazonov commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
Two typos in the line.
```suggestion
During `Import-Module` a module will be loaded into a separate Windows PowerShell process instead of current PowerShell Core if:<br />
```

anmenaga replied:
fixed.

iSazonov commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
We could keep the remoting for perf reasons until force module unload or current pwsh session exit.

anmenaga replied:
True. On the other hand, having a WinPS process + remoting session hanging around even if a cmdlet was used only once/few times is also a waste of resources. After sync with @SteveL-MSFT this can change if such user feedback comes in.

iSazonov commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
If serialization and proxing is key in the proposal why do we limit ourselves only Windows? I'd expect that this works from Unix too.

anmenaga replied:
I don't quite understand your thoughts here about Unix.
The main problem that this work is trying to solve - making it easier to use (some) WindowsPS-only modules on PS Core on windows. So the entire problem statement is restricted to Windows only.

iSazonov replied:
If serialization and proxing is used it could work from Unix client too.

anmenaga replied:
Oh. Yes, theoretically it can (however not with the PSJob's transport).
It would have been an interesting scenario: using WindowsPS-only modules on a remote Windows machine from PS Core on Linux. :)

iSazonov commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
There should be runspace afinity between these processes, otherwise some runspaces from Core will communicate with one runspace in Windows PowerShell that it wrong I believe.

anmenaga replied:
Valid point; In some cases it is fine; in others - it is not. Unfortunately remoting transport currently does not allow this and implementing this is a significant amount of work. I'll add your point to the Alternate Considerations to keep track of this.

KirkMunro commented:
What about providers? How will this enable working with a custom PSDrive that gets created when an incompatible module is loaded in the remote session? The WebAdministration module that ships with IIS on Windows 10 is a good example of a module that needs interaction with the custom provider via core cmdlets to perform automation properly.

joeyaiello commented:
Couple questions/comments from @PowerShell/powershell-committee today:
* @daxian-dbw wondered why we didn't use the same transport leveraged by PSJobs (where the `powershell.exe` process is launched in "server mode") vs. named pipes
* @JamesWTruher raised that the user is likely not going to be expected deserialized objects and it could be difficult to debug or understand what's going on in long-running tasks. While maybe not enough, we agree that we should definitely throw a warning that `powershell.exe` is going to be used, that a deserialized object will be emitted, and that `Import-Module -SkipEditionCheck` can be used if you know what you're doing to explicitly load the module into the PowerShell Core. 
* Similarly, we should do this for auto-loading of modules at cmdlet execution time.
* @daxian-dbw also raised that the debug channel is likely occupied by this as long as the `powershell` process is open (which is the case so long as any Windows PS modules are imported). To me, this is a bit of a showstopper, and we should investigate the jobs channel to see if that could alleviate our problem here.
* We should validate that, given that the proxy module is the same name as the original module, our new module telemetry reports that the proxy module was imported if it's in our first-party allow list

anmenaga commented:
* Added `-UseWindowsPowerShell` parameter to allow explicit use of this feature for modules outside of system32
* providers are not covered by this. Added this point to Alternate Considerations. In some cases, doing the entire scenario, that can not be split on remoting boundaries, is possible since the WinCompat remoting session is available `Invoke-Command -Session WinPSCompatSession -ScriptBlock {Get-WinPSLiveObject | Set-WinPSLiveObject}`
* PS transport was changed to the one used by PSJobs; Valid point about IPC channel blocking debugging of WinPS.
* added a warning to be generated every time a module is imported using this new functionality
* added that new functionality should also be supported during module autoload / command discovery.
* added new telemetry type.

anmenaga commented:
Added 'Module blacklist' section.

joeyaiello commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
Sounds like this one got fixed

joeyaiello commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
Doesn't Hyper-V already work natively? Or is there something here because of the SxS Hyper-V module versions?

joeyaiello commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
As of today, this hasn't been implemented

SteveL-MSFT replied:
This is now implemented.

joeyaiello commented on 2-Draft-Accepted/RFCNNN-Importing-Windows-PowerShell-modules-in-PowerShell-Core.md:
```suggestion

Going forward, we will continue to update this list if new modules are discovered to have problems operating in this compatibility layer.
```

aolszowka commented:
Now that this PR is merged and the legacy project closed; if we're finding issues with the capability layer (such as https://github.com/PowerShell/WindowsCompatibility/issues/89)  should they be relocated to the main project?
